File: src/main/java/com/portofino/realtrainmodunofficial/block/SignalRemoteBlock.java
package com.portofino.realtrainmodunofficial.block;

import com.mojang.serialization.MapCodec;
import com.portofino.realtrainmodunofficial.ClientHooks;
import com.portofino.realtrainmodunofficial.RealTrainModUnofficialBlockEntities;
import com.portofino.realtrainmodunofficial.blockentity.SignalRemoteBlockEntity;
import net.minecraft.core.BlockPos;
import net.minecraft.world.ItemInteractionResult;
import net.minecraft.world.InteractionResult;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.level.BlockGetter;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.BaseEntityBlock;
import net.minecraft.world.level.block.RenderShape;
import net.minecraft.world.level.block.SoundType;
import net.minecraft.world.level.block.entity.BlockEntity;
import net.minecraft.world.level.block.entity.BlockEntityTicker;
import net.minecraft.world.level.block.entity.BlockEntityType;
import net.minecraft.world.level.block.state.BlockBehaviour;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.phys.BlockHitResult;
import net.minecraft.world.phys.shapes.CollisionContext;
import net.minecraft.world.phys.shapes.VoxelShape;
import org.jetbrains.annotations.Nullable;

/**
 * 受信機と変更機は見た目と役割だけが違うので、同じブロック実装を共有します。
 */
public class SignalRemoteBlock extends BaseEntityBlock {
    public enum Mode {
        RECEIVER,
        CHANGER,
        VALUE_INPUT
    }

    private final Mode mode;
    private final MapCodec<SignalRemoteBlock> codec;

    public SignalRemoteBlock(BlockBehaviour.Properties properties, Mode mode) {
        super(properties);
        this.mode = mode;
        this.codec = simpleCodec(blockProperties -> new SignalRemoteBlock(blockProperties, this.mode));
    }

    public SignalRemoteBlock(Mode mode) {
        this(BlockBehaviour.Properties.of().sound(SoundType.METAL).strength(1.0F, 6.0F), mode);
    }

    @Override
    protected MapCodec<? extends BaseEntityBlock> codec() {
        return codec;
    }

    @Override
    public RenderShape getRenderShape(BlockState state) {
        return RenderShape.MODEL;
    }

    @Override
    public VoxelShape getShape(BlockState state, BlockGetter level, BlockPos pos, CollisionContext context) {
        return box(1.0D, 0.0D, 1.0D, 15.0D, 14.0D, 15.0D);
    }

    @Override
    protected VoxelShape getCollisionShape(BlockState state, BlockGetter level, BlockPos pos, CollisionContext context) {
        return getShape(state, level, pos, context);
    }

    @Nullable
    @Override
    public BlockEntity newBlockEntity(BlockPos pos, BlockState state) {
        return new SignalRemoteBlockEntity(pos, state);
    }

    @Override
    public <T extends BlockEntity> BlockEntityTicker<T> getTicker(Level level, BlockState state, BlockEntityType<T> type) {
        return null;
    }

    @Override
    protected ItemInteractionResult useItemOn(ItemStack stack, BlockState state, Level level, BlockPos pos, Player player,
                                              net.minecraft.world.InteractionHand hand, BlockHitResult hit) {
        openScreen(level, pos);
        return ItemInteractionResult.sidedSuccess(level.isClientSide);
    }

    @Override
    protected InteractionResult useWithoutItem(BlockState state, Level level, BlockPos pos, Player player, BlockHitResult hit) {
        openScreen(level, pos);
        return InteractionResult.sidedSuccess(level.isClientSide);
    }

    private void openScreen(Level level, BlockPos pos) {
        if (!level.isClientSide) {
            return;
        }
        switch (mode) {
            case CHANGER -> ClientHooks.openSignalChangerScreen(pos);
            case VALUE_INPUT -> ClientHooks.openSignalValueScreen(pos);
            case RECEIVER -> {
                if (level.getBlockEntity(pos) instanceof SignalRemoteBlockEntity blockEntity
                    && blockEntity.getLinkedChannel() > 0) {
                    ClientHooks.openSignalChangerScreen(pos);
                } else {
                    ClientHooks.openSignalReceiverScreen(pos);
                }
            }
        }
    }
}

File: src/main/kotlin/cc/mirukuneko/realtrainmodrenewed/block/SignalRemoteBlock.kt
// SPDX-License-Identifier: LGPL-3.0-or-later
// Copyright © 2026 mirukuneko and RealTrainModRenewed contributors
package cc.mirukuneko.realtrainmodrenewed.block

import com.mojang.serialization.MapCodec
import cc.mirukuneko.realtrainmodrenewed.RealTrainModRenewedBlockEntities
import cc.mirukuneko.realtrainmodrenewed.blockentity.SignalRemoteBlockEntity
import net.minecraft.core.BlockPos
import net.minecraft.server.level.ServerLevel
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.InteractionHand
import net.minecraft.world.InteractionResult
import net.minecraft.world.entity.player.Player
import net.minecraft.world.level.Level
import net.minecraft.world.level.block.BaseEntityBlock
import net.minecraft.world.level.block.Block
import net.minecraft.world.level.block.RenderShape
import net.minecraft.world.level.block.entity.BlockEntity
import net.minecraft.world.level.block.entity.BlockEntityTicker
import net.minecraft.world.level.block.entity.BlockEntityType
import net.minecraft.world.level.block.state.BlockBehaviour
import net.minecraft.world.level.block.state.BlockState
import net.minecraft.world.phys.BlockHitResult
import net.minecraft.world.phys.shapes.CollisionContext
import net.minecraft.world.phys.shapes.VoxelShape

class SignalRemoteBlock(properties: BlockBehaviour.Properties, val mode: Mode) : BaseEntityBlock(properties) {
    enum class Mode { RECEIVER, CHANGER, VALUE_INPUT }

    companion object {
        private val CODEC_RECEIVER: MapCodec<SignalRemoteBlock> = simpleCodec { props -> SignalRemoteBlock(props, Mode.RECEIVER) }
        private val CODEC_CHANGER: MapCodec<SignalRemoteBlock> = simpleCodec { props -> SignalRemoteBlock(props, Mode.CHANGER) }
        private val CODEC_VALUE_INPUT: MapCodec<SignalRemoteBlock> = simpleCodec { props -> SignalRemoteBlock(props, Mode.VALUE_INPUT) }
    }

    override fun codec(): MapCodec<out BaseEntityBlock> = when (mode) {
        Mode.RECEIVER -> CODEC_RECEIVER
        Mode.CHANGER -> CODEC_CHANGER
        Mode.VALUE_INPUT -> CODEC_VALUE_INPUT
    }

    override fun newBlockEntity(pos: BlockPos, state: BlockState): BlockEntity =
        SignalRemoteBlockEntity(pos, state)

    override fun getRenderShape(state: BlockState): RenderShape = RenderShape.MODEL

    override fun getShape(state: BlockState, level: net.minecraft.world.level.BlockGetter, pos: BlockPos, ctx: CollisionContext): VoxelShape =
        net.minecraft.world.phys.shapes.Shapes.block()

    override fun <T : BlockEntity> getTicker(level: Level, state: BlockState, type: BlockEntityType<T>): BlockEntityTicker<T>? =
        null

    override fun useWithoutItem(state: BlockState, level: Level, pos: BlockPos, player: Player, hit: BlockHitResult): InteractionResult {
        if (level.isClientSide) return InteractionResult.SUCCESS
        val be = level.getBlockEntity(pos) as? SignalRemoteBlockEntity ?: return InteractionResult.FAIL
        if (player is ServerPlayer) {
            val channel = be.linkedChannel
            player.sendSystemMessage(net.minecraft.network.chat.Component.literal("Channel: $channel"))
        }
        return InteractionResult.SUCCESS
    }
}
